### ISSUE TITLE
Ensure all runs have a concrete channel label

### ISSUE THREAD

**Issue by foolip** [web-platform-tests/wpt.fyi#1539]
Iterating all 16,189 current master runs using wpt.fyi/api/runs, 565 of them were found to not have a channel representing a concrete channel, that is no label in `['stable', 'beta', 'dev', 'canary', 'nightly', 'preview']`. [sans-channel.json](https://gist.github.com/foolip/b5bde85ea6841fd001c5486727ebf1be) is a JSON list of these. 

There are 150 Chrome runs, 1 Edge run, 149 Firefox runs and 265 Safari runs. All of them have the experimental label.

Looking at the `browser_version` values, it's clear that the Chrome runs should have the 'dev' label and Safari runs the 'preview' label.

The Firefox runs are all from Buildbot and have versions '66.0a1', '67.0a1' or '68.0a1'. I'm almost certain they're Firefox Nightly runs, but would ask @jugglinmike to confirm.

The single Edge run is one of very few in https://wpt.fyi/runs?label=msedge, from a setup that never became viable. Still, the correct label for the experimental ones there is 'preview', as in Window Insider Preview.

I came across this issue while trying to make this code differentiate between Firefox Dev and Nightly:
https://github.com/web-platform-tests/wpt.fyi/blob/4a1621ee74a070d320a7b855eb902fbe7b3704ec/webapp/components/display-logo.js#L121-L129

By not handling 'experimental' there at all, the mapping wouldn't get it wrong for any of these 565 existing runs.

**Comment by foolip** [comment]
@Hexcles do you still turn to https://github.com/web-platform-tests/results-analysis to fix up things like this?

**Comment by foolip** [comment]
I've sent https://github.com/web-platform-tests/wpt.fyi/pull/1541 to remove the `name.replace(/-experimental$/, '')` that didn't do anything in that code snipped.

**Comment by jugglinmike** [comment]
[We began collecting results from Firefox Nightly on April 11, 2018.](https://github.com/web-platform-tests/results-collection/pull/538) Any results submitted prior to this date must be for the stable release channel. Results submitted after this date could be from either channel.

I can't find documentation about Firefox's version specifiers, but [judging from the recent release history](https://en.wikipedia.org/wiki/Firefox_version_history), the trailing "a" seems to indicate a Nightly build. @jgraham can you verify that?

**Comment by jgraham** [comment]
I believe that's true, but only based on observation.

**Comment by foolip** [comment]
The earliest of the Firefox runs in [sans-channel.json](https://gist.github.com/foolip/b5bde85ea6841fd001c5486727ebf1be) is https://wpt.fyi/results/?run_id=6611871985565696&run_id=6232870918029312, here shown next to the stable run from Buildbot. Beta at the time was 65 and doesn't have an "a" (or "b") in the version.

So I think we can safely label all the Firefox results as "nightly".

**Comment by stephenmcgruer** [comment]
Ok, I have a WIP fixup that addresses most of this. However it turns out that there are also non-preview Safari entries in there, e.g.:

```
  {
    "id": 6210255230861312,
    "browser_name": "safari",
    "browser_version": "12.1",
    "os_name": "macos",
    "os_version": "10.13",
    "revision": "4c7096bac0",
    "full_revision_hash": "4c7096bac0c8e8b6254d54eb3b1200c4ad2e93fe",
    "results_url": "https://storage.googleapis.com/wptd/4c7096bac0c8e8b6254d54eb3b1200c4ad2e93fe/safari-12.1-macos-10.13-7624afe615-summary.json.gz",
    "created_at": "2019-01-05T10:09:33.864144Z",
    "time_start": "2019-01-05T00:12:16.423Z",
    "time_end": "2019-01-05T09:54:02.159Z",
    "raw_results_url": "https://storage.googleapis.com/wptd-results/4c7096bac0c8e8b6254d54eb3b1200c4ad2e93fe/safari-12.1-macos-10.13-7624afe615/report.json",
    "labels": [
      "buildbot",
      "experimental",
      "master",
      "safari"
    ]
  },
```

Are those just stable?

**Comment by stephenmcgruer** [comment]
Actually, that Safari run is marked `experimental`. In https://github.com/web-platform-tests/wpt.fyi/tree/master/api/README.md we describe `experimental` as a release channel. What's the real story here - what does experimental mean? @foolip

**Comment by stephenmcgruer** [comment]
Looking at https://wpt.fyi/runs?label=master, I find

```
Chrome:
  chrome, dev, experimental, master, taskcluster, user:foo
  chrome, master, stable, taskcluster, user:foo
Edge:
  azure, canary, edge, edgechromium, master
  azure, dev, edge, edgechromium, experimental, master
  buildbot, edge, master, stable
Firefox:
  experimental, firefox, master, nightly, taskcluster, user:foo
  firefox, master, stable, taskcluster, user:foo
Safari:
  azure, experimental, master, preview, safari
  azure, master, safari, stable
```

So...

1. `experimental` is not a release channel...?
1. but `stable` and `experimental` should also not co-exist
1. also Edge Canary *doesn't* have `experimental` ?

Also, a bunch of the code in WPT assumes `experimental` is a release channel. Perhaps just to deal with stuff like this, or perhaps not. 

What is the source of truth for release channels and labels?

**Comment by Hexcles** [comment]
Late to the part, I fixed the bulk of this case last time using https://github.com/web-platform-tests/data-migration/blob/master/tagger/channels.go

I haven't looked into why these ~500 runs were missed.

**Comment by stephenmcgruer** [comment]
Right, and looks like https://github.com/web-platform-tests/data-migration/blob/master/tagger/stable.go was designed to fix the 'you should be stable, not experimental' problem.

Ok, let me try to use the existing code to see if I can find + fixup the same cases.

**Comment by jgraham** [comment]
I wasn under the impression that wpt.fyi was using `experimental` to mean "whichever the most unstable release channel we run is". For Firefox that's "nightly", for Chrome "dev" and for Safari "preview". Since each browser is using different terminology there's some attempt to smooth over the differences.

**Comment by Hexcles** [comment]
Prior discussion: https://github.com/web-platform-tests/wpt.fyi/issues/538

**Comment by stephenmcgruer** [comment]
I've run the `channel.go` fix-up on both staging + prod, which should have tagged the preview/dev/nightly builds ([prod.log](https://github.com/web-platform-tests/wpt.fyi/files/3826184/prod.log) in case I screwed up). Next up, I will be running the stable fix-up.

**Comment by stephenmcgruer** [comment]
Most of the stable fix-up has finished (
[prod-stable-rewrite.log](https://github.com/web-platform-tests/wpt.fyi/files/3826451/prod-stable-rewrite.log)), but I ran into a fun problem:

```
panic: datastore: cannot load field "full_revision_hash" into a "shared.TestRun": no such struct field

goroutine 86378 [running]:
_/usr/local/google/home/smcgruer/github/data-migration/processor.ProcessRun(0xce4d60, 0xc0000340b8, 0xcd9e60, 0x11fbac0, 0xc000329350, 0xc0012e5380, 0xc00034b130)
	/usr/local/google/home/smcgruer/github/data-migration/processor/migrate.go:46 +0x2e5
created by _/usr/local/google/home/smcgruer/github/data-migration/processor.MigrateData
	/usr/local/google/home/smcgruer/github/data-migration/processor/migrate.go:87 +0x2f3
exit status 2
```

Apparently something is missing some data :d

Checking the original generation code for runs missing concrete channel labels, we're down to 4 on staging and 10 on prod.

**Comment by Hexcles** [comment]
The field name in Datastore should be `FullRevisionHash`.

Datastore, being schema-less, allows you to add any field. It's possible this run was created somehow manually.

**Comment by stephenmcgruer** [comment]
I believe its 5768854936485888 . https://wpt.fyi/api/runs/5768854936485888 is fun :D

**Comment by foolip** [comment]
@jgraham yes the experimental label means something like most bleeding edge configuration available at the time of running. Well, except for Edge Canary I guess, we label Dev as experimental still.

With the real channel name always present as a label it’ll be easier to fix the UI to list the real channel names (both Edge Dev and Canary) and write queries that will always stick to a single channel.

**Comment by stephenmcgruer** [comment]
I touched up the script to deal with the error above, and we're down to the single `msedge` run on prod and two edge runs on staging. Nearly there :)

**Comment by stephenmcgruer** [comment]
And we should be done. All runs (at least master ones) should have a proper channel on both staging and prod.

Do we have an issue for the fact that Edge Canary doesn't seem to be tagged with `experimental`? (Or is there a valid reason it shouldn't be).

**Comment by foolip** [comment]
There’s no issue for that AFAIK, and I don’t recall how it came about. @mustjab do you know why it’s Edge Dev and not Canary that shows on wpt.fyi by default?

**Comment by stephenmcgruer** [comment]
The code reason, incidentally, would seem to be that `results-processor/wptreport.py` has a `CHANNEL_TO_LABEL` map that doesn't include 'canary' (it maps ['release', 'stable'] --> 'stable', ['beta'] --> 'beta', and ['dev', 'experimental', 'nightly', 'preview'] --> 'experimental'), and that the report from azure doesn't include 'experimental' either.

**Comment by foolip** [comment]
I've confirmed using my original method that no more runs remain without a concrete channel. Here's the channel map I used:
```
  const channels = {
    'chrome': ['stable', 'beta', 'dev', 'canary'],
    'edge': ['stable', 'beta', 'dev', 'canary', 'preview'],
    'firefox': ['stable', 'beta', 'dev', 'nightly'],
    'safari': ['stable', 'preview'],
    'webkitgtk': ['stable', 'nightly'],
  };
```

I didn't check that there were runs for each combination, just that nothing fell outside these expected channels. Note that [Edge preview](https://wpt.fyi/runs?product=edge&label=preview) is a rare setup we had on Windows Insider Preview briefly before Chromium-based Edge was announced. And we don't have any Edge beta runs yet, but probably will.

@stephenmcgruer the above channel lists would be a decent start to fix https://github.com/web-platform-tests/wpt.fyi/issues/1577.
